src/functions/functionsDialog.component.tsx

```diff
@@ -0,0 +1,363 @@
+        const updatedErrorState: FunctionErrorState[string] = {
+          ...errors[id],
+          ...value, // Spread the value object to update name and/or expression
+        };
+
+        // Replace the old error state with the updated one
+        if (value) {
+          newErrors[id] = updatedErrorState;
+        } else {
+          delete newErrors[id];
+        }
+
+        return newErrors;
+      }),
+    []
+  );
+  React.useEffect(() => {
+    setFunctions(
+      appliedFunctions.length === 0
+        ? [
+            {
+              id: crypto.randomUUID(),
+              name: '',
+              expression: [],
+              dataType: 'scalar',
+              channels: [],
+            },
+          ]
+        : appliedFunctions
+    );
+    setErrors({});
+  }, [appliedFunctions]);
+
+  const handleError = React.useCallback(
+    (error: AxiosError) => {
+      const errorCode = (error.response?.data as APIError).detail;
+
+      if (typeof errorCode === 'string' && !errorCode.includes(':')) return;
+      const parsedErrors = parseErrorCode(errorCode);
+      parsedErrors.forEach((error) => {
+        const { index, errorMessage, isNameError } = error;
+        const functionId = functions[index].id;
+        handleChangeError(functionId)(
+          isNameError
+            ? { name: { message: errorMessage } }
+            : { expression: { message: errorMessage } }
+        );
+      });
+    },
+    [functions, handleChangeError]
+  );
+
+  const { mutateAsync: postValidateFunctions } = usePostValidateFunctions();
+
+  const checkErrors = React.useCallback(
+    (index: number, id: string) => {
+      const functionsToValidate = functions.slice(0, index + 1);
+      postValidateFunctions(functionsToValidate)
+        .then(() => {
+          setErrors((errors) => {
+            const updatedErrors = JSON.parse(JSON.stringify(errors));
+            delete updatedErrors[id];
+            return updatedErrors;
+          });
+        })
+        .catch(handleError);
+    },
+    [functions, handleError, postValidateFunctions]
+  );
+
+  const applyFunctions = React.useCallback(
+    (newFunctions: ValidateFunctionState[]) => {
+      postValidateFunctions(functions)
+        .then((response: string[]) => {
+          dispatch(
+            changeAppliedFunctions(
+              newFunctions.map((func, index) => ({
+                ...func,
+                dataType: response[index] as DataType,
+              }))
+            )
+          );
+          dispatch(
+            updateSelectedColumns([
+              ...appliedSelectedIds,
+              ...newFunctions.map((func) => func.name),
+            ])
+          );
+          onClose();
+        })
+        .catch(handleError);
+    },
+    [
+      postValidateFunctions,
+      functions,
+      handleError,
+      dispatch,
+      appliedSelectedIds,
+      onClose,
+    ]
+  );
+
+  const functionsToken: FunctionTag[] = functions
```

Comment:
Confusingly named - maybe `tokenisedFunctions`? Made even worse with the token/tag crossover here